packages/ifpack2/src/Ifpack2_BlockTriDiContainer_impl.hpp

```diff
@@ -527,16 +607,246 @@ namespace Ifpack2 {
+        const local_ordinal_type num_vectors = mv.extent(1);
+        const local_ordinal_type mv_blocksize = blocksize*num_vectors;
+
+        // 0. post receive async
+        for (local_ordinal_type i=0,iend=pids.recv.extent(0);i<iend;++i) {
+          irecv(comm, 
+                reinterpret_cast<char*>(buffer.recv.data() + offset_host.recv[i]*mv_blocksize),
+                (offset_host.recv[i+1] - offset_host.recv[i])*mv_blocksize*sizeof(impl_scalar_type),
+                pids.recv[i],
+                42,
+                &reqs.recv[i]);
+        }
+
+        // 1. async memcpy
+#if defined (KOKKOS_ENABLE_OPENMP)
+#pragma omp parallel for
+#endif        
+        for (local_ordinal_type i=0;i<static_cast<local_ordinal_type>(pids.send.extent(0));++i) {
+          auto &stream_at_i = stream.send[i];
+          // 1.0. enqueue pack buffer
+          copyViaCudaStream<ToBuffer>(lids.send, buffer.send, 
+                                      offset_host.send(i), offset_host.send(i+1), 
+                                      mv, blocksize,
+                                      stream_at_i);
+          
+          // // 1.1. enqueue call back posting isend , call back does not really work; what did I do wrong.
+          // auto &data_at_i = callback_data.send[i];
+          // data_at_i.comm = comm;
+          // data_at_i.buf = (char*)(buffer.send.data() + offset_host.send[i]*mv_blocksize);
+          // data_at_i.count = (offset_host.send[i+1] - offset_host.send[i])*mv_blocksize*sizeof(impl_scalar_type);
+          // data_at_i.pid = pids.send[i];
+          // data_at_i.tag = 42;
+          // data_at_i.ireq = &reqs.send[i];
+          // data_at_i.r_val = 0;
+          // CUDA_SAFE_CALL(cudaStreamAddCallback(stream_at_i, callback_isend, (void*)&data_at_i, 0));
+        }
+
+        Kokkos::fence();
+#if defined (KOKKOS_ENABLE_OPENMP)
+#pragma omp parallel for
+#endif        
+        for (local_ordinal_type i=0;i<static_cast<local_ordinal_type>(pids.send.extent(0));++i) {
+          // 1.1. sync the stream and isend
+          //CUDA_SAFE_CALL(cudaStreamSynchronize(stream_at_i));
+          isend(comm, 
+                reinterpret_cast<const char*>(buffer.send.data() + offset_host.send[i]*mv_blocksize),
+                (offset_host.send[i+1] - offset_host.send[i])*mv_blocksize*sizeof(impl_scalar_type),
+                pids.send[i], 
+                42,
+                &reqs.send[i]);
+        }
+        
+        // 2. poke communication
+        for (local_ordinal_type i=0,iend=pids.recv.extent(0);i<iend;++i) {
+          int flag;
+          MPI_Status stat;
+          MPI_Iprobe(pids.recv[i], 42, comm, &flag, &stat);
+        }
+#endif // HAVE_IFPACK2_MPI
+      }
+
+      void syncRecvViaCudaStream() {
+	IFPACK2_BLOCKTRIDICONTAINER_TIMER("BlockTriDi::AsyncableImport::SyncRecv");
+#ifdef HAVE_IFPACK2_MPI
+        // 0. wait for receive async.
+#if defined (KOKKOS_ENABLE_OPENMP)
+#pragma omp parallel for
```

Comment:
This parallel loop requires `MPI_THREAD_MULTIPLE`.  I worry that we don't have adequate testing for the case where MPI does not provide `MPI_THREAD_MULTIPLE`.

Reply:
Resolved -- see above.